src/persist-client/src/impl/machine.rs

```diff
@@ -719,15 +719,18 @@ mod tests {
     use crate::async_runtime::CpuHeavyRuntime;
     use crate::batch::{validate_truncate_batch, BatchBuilder};
     use crate::r#impl::compact::{CompactReq, Compactor};
+    use crate::r#impl::paths::PartialBlobKey;
+    use crate::r#impl::state::ProtoStateRollup;
     use crate::read::{fetch_batch_part, Listen, ListenEvent};
     use crate::tests::new_test_client;
-    use crate::{PersistConfig, ShardId};
+    use crate::{GarbageCollector, PersistConfig, ShardId};
 
     use super::*;
 
     #[derive(Debug, Default)]
     struct DatadrivenState {
         batches: HashMap<String, HollowBatch<u64>>,
+        batch_keys_to_names: HashMap<Vec<PartialBlobKey>, String>,
```

Comment:
instead of maintaining the denormalization here, thoughts on iterating though `batches` below to find the one with a given set of keys?  it's pretty awful from a big O perspective, but things should always be pretty low cardinality